File: client/web/src/services/bankService.ts
import { apiService } from '../core/api';

export interface QuestionBank {
  _id: string;
  name: string;
  description?: string | null;
  type: 'personal' | 'school';
  schoolId?: string | null;
  createdBy: string | { _id: string; name?: string };
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface BankMember {
  _id: string;
  bankId: string;
  userId: string | { _id: string; name: string; email?: string; role?: string };
  role: 'owner' | 'manager' | 'viewer';
  status: 'active' | 'pending';
  invitedBy?: string;
  invitedAt?: string;
  approvedBy?: string;
  approvedAt?: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface BankSummary {
  bank: QuestionBank;
  membership: {
    role: 'owner' | 'manager' | 'viewer';
    status: 'active' | 'pending';
  } | null;
}

export const bankService = {
  listBanks: () =>
    apiService.get<QuestionBank[]>('/banks'),

  createBank: (data: {
    name: string;
    description?: string;
    type?: 'personal' | 'school';
    schoolId?: string;
  }) => apiService.post<QuestionBank>('/banks', data),

  getBank: (bankId: string) =>
    apiService.get<{ bank: QuestionBank; membership: BankMember | null }>(`/banks/${bankId}`),

  listMembers: (bankId: string, status?: 'active' | 'pending') =>
    apiService.get<{ results: BankMember[] }>(`/banks/${bankId}/members`, {
      params: status ? { status } : undefined,
    }),

  inviteMember: (bankId: string, userId: string) =>
    apiService.post<BankMember>(`/banks/${bankId}/members`, { userId }),

  updateMemberRole: (bankId: string, userId: string, role: 'manager' | 'viewer') =>
    apiService.patch<BankMember>(`/banks/${bankId}/members/${userId}`, { role }),

  removeMember: (bankId: string, userId: string) =>
    apiService.delete<void>(`/banks/${bankId}/members/${userId}`),

  leaveBank: (bankId: string) =>
    apiService.post<void>(`/banks/${bankId}/leave`),

  requestAccess: (bankId: string) =>
    apiService.post<BankMember>(`/banks/${bankId}/request-access`),

  listPendingRequests: (bankId: string) =>
    apiService.get<{ results: BankMember[] }>(`/banks/${bankId}/requests/pending`),

  respondToRequest: (
    bankId: string,
    userId: string,
    decision: 'approve' | 'reject'
  ) => apiService.post<BankMember | { status: string }>(
    `/banks/${bankId}/requests/${userId}/respond`,
    { decision }
  ),

  transferOwnership: (bankId: string, toUserId: string) =>
    apiService.post<BankMember>(`/banks/${bankId}/transfer-ownership`, { toUserId }),
};

export const notificationService = {
  getUnreadCount: () =>
    apiService.get<{ unreadCount: number }>('/notifications/unread-count'),

  list: (params?: { isRead?: boolean; type?: string; page?: number; limit?: number }) =>
    apiService.get<{
      results: Array<{
        _id: string;
        userId: string;
        type: string;
        title: string;
        body: string;
        data?: Record<string, unknown>;
        isRead: boolean;
        readAt?: string;
        createdAt: string;
      }>;
      page: number;
      limit: number;
      total: number;
      pages: number;
    }>('/notifications', { params: params as Record<string, string | number | boolean> }),

  markAsRead: (id: string) =>
    apiService.post<unknown>(`/notifications/${id}/read`),

  markAllAsRead: () =>
    apiService.patch<void>('/notifications/read-all'),
};

File: client/mobile/lib/core/network/bank_service.dart
import '../../core/network/api_client.dart';
import '../../core/constants/app_constants.dart';
import '../../domain/entities/question_bank.entity.dart';
import '../../domain/entities/bank_membership.entity.dart';

class BankDetail {
  final QuestionBank bank;
  final BankMembership? membership;

  BankDetail({
    required this.bank,
    this.membership,
  });

  factory BankDetail.fromJson(Map<String, dynamic> json) {
    return BankDetail(
      bank: QuestionBank.fromJson(json['bank'] as Map<String, dynamic>),
      membership: json['membership'] != null
          ? BankMembership.fromJson(json['membership'] as Map<String, dynamic>)
          : null,
    );
  }
}

class PaginatedBanks {
  final List<QuestionBank> results;
  final int total;
  final int page;
  final int pages;

  PaginatedBanks({
    required this.results,
    required this.total,
    required this.page,
    required this.pages,
  });

  factory PaginatedBanks.fromJson(Map<String, dynamic> json) {
    return PaginatedBanks(
      results: (json['results'] as List<dynamic>)
          .map((e) => QuestionBank.fromJson(e as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int? ?? 0,
      page: json['page'] as int? ?? 1,
      pages: json['pages'] as int? ?? 1,
    );
  }
}

class BankService {
  BankService({required ApiClient apiClient}) : _apiClient = apiClient;

  final ApiClient _apiClient;

  /// GET /banks → returns flat QuestionBank[]
  Future<List<QuestionBank>> listBanks() {
    return _apiClient.get<List<QuestionBank>>(
      ApiConstants.banks,
      parser: (data) {
        final list = data as List<dynamic>;
        return list
            .map((e) => QuestionBank.fromJson(e as Map<String, dynamic>))
            .toList();
      },
    );
  }

  /// GET /banks/:bankId → returns { bank, membership }
  Future<BankDetail> getBank(String bankId) {
    return _apiClient.get<BankDetail>(
      '${ApiConstants.banks}/$bankId',
      parser: (data) => BankDetail.fromJson(data as Map<String, dynamic>),
    );
  }

  /// POST /banks → returns flat QuestionBank
  Future<QuestionBank> createBank({
    required String name,
    String? description,
    required String type,
  }) {
    return _apiClient.post<QuestionBank>(
      ApiConstants.banks,
      data: {
        'name': name,
        if (description != null && description.isNotEmpty) 'description': description,
        'type': type,
      },
      parser: (data) => QuestionBank.fromJson(data as Map<String, dynamic>),
    );
  }

  /// GET /banks/search → returns { results, total, page, pages }
  Future<PaginatedBanks> searchBanks(String query) {
    return _apiClient.get<PaginatedBanks>(
      '${ApiConstants.banks}/search',
      queryParameters: {'q': query},
      parser: (data) => PaginatedBanks.fromJson(data as Map<String, dynamic>),
    );
  }

  /// POST /banks/:bankId/request-access → returns flat BankMembership
  Future<BankMembership> requestAccess(String bankId) {
    return _apiClient.post<BankMembership>(
      '${ApiConstants.banks}/$bankId/request-access',
      parser: (data) => BankMembership.fromJson(data as Map<String, dynamic>),
    );
  }
}
